File: homework4/design_pattern/kotlin/chain.kt
class Calculator {
    private var result: Double = 0.0

    fun add(value: Double): Calculator {
        result += value
        return this // Return itself
    }

    fun subtract(value: Double): Calculator {
        result -= value
        return this // Return itself
    }

    fun multiply(value: Double): Calculator {
        result *= value
        return this // Return itself
    }

    fun divide(value: Double): Calculator {
        if (value != 0.0) {
            result /= value
        }
        return this // Return itself
    }

    fun getResult(): Double {
        return result
    }
}

fun main() {
    val calc = Calculator()
    val result = calc.add(5.0)
        .multiply(2.0)
        .subtract(3.0)
        .divide(2.0)
        .getResult()
    println(result)
}

File: homework4/design_pattern/csharp/chain.cs
using System;

class Calculator
{
    private double result;

    public Calculator()
    {
        result = 0;
    }

    public Calculator Add(double value)
    {
        result += value;
        return this; // Return itself for method chaining
    }

    public Calculator Subtract(double value)
    {
        result -= value;
        return this; // Return itself for method chaining
    }

    public Calculator Multiply(double value)
    {
        result *= value;
        return this; // Return itself for method chaining
    }

    public Calculator Divide(double value)
    {
        if (value != 0)
        {
            result /= value;
        }
        return this; // Return itself for method chaining
    }

    public double GetResult()
    {
        return result;
    }
}

class Program
{
    static void Main(string[] args)
    {
        Calculator calc = new Calculator();
        double result = calc.Add(5)
                            .Multiply(2)
                            .Subtract(3)
                            .Divide(2)
                            .GetResult();
        Console.WriteLine(result);
    }
}